src/main/scala/advent/solutions/Day3.scala

```diff
@@ -48,8 +52,156 @@ object Day3 {
     *
     * @param text A comma separated string of displacements e.g "R75,D30,R83"
     */
-  def parse(text: String): Either[WireParseError, Wire] = {
-    ???
+  def parse(text: String): Either[List[WireParseError], Wire] = {
+    val texts: List[(String, String)] =
+      text
+        .split(",")
+        .map(t => (t.filter(_.isLetter), t.filter(!_.isLetter)))
+        .toList
+
+    val (lefts, rights) = checkTexts(texts).partitionMap(identity)
+
+    if (lefts.isEmpty) Right(Wire(rights)) else Left(lefts)
+  }
+
+  private def checkTexts(
+      texts: List[(String, String)]
+  ): List[Either[WireParseError, Displacement]] = {
+    val out = texts.map(t =>
+      validateDirection(t._1) match {
+        case Right(direction) =>
+          validateDistance(t._2) match {
+            case Right(distance) =>
+              Right(Displacement(direction, distance.toInt))
+            case Left(value) => Left(value)
+          }
+        case Left(value) => Left(value)
+      }
+    )
+    out
+  }
+
+  private def validateDirection(
+      s: String
+  ): Either[InvalidDirection, Direction] = {
+    if (s == "L" | s == "R" | s == "U" | s == "D") {
+      s match {
+        case "L" => Right(Direction.Left)
+        case "R" => Right(Direction.Right)
+        case "U" => Right(Direction.Up)
+        case "D" => Right(Direction.Down)
+      }
+    } else Left(InvalidDirection(s))
+  }
+
+  private def validateDistance(s: String): Either[InvalidDistance, String] = {
+    if (s.toInt > 0) Right(s) else Left(InvalidDistance(s))
+  }
+
+  final case class Coordinate(x: Int, y: Int)
+
+  final case class StartingAndEndingPoints(start: Coordinate, end: Coordinate)
+
+  final case class Path(points: Seq[Coordinate])
+
+  final case class WireMap(
+      displacement: Displacement,
+      startingAndEndingPoints: StartingAndEndingPoints
+  )
+
+  final case class PathLocations(displacement: Displacement, path: Path)
+
+  import Direction._
+
+  /**
+    *
+    * @param wire
+    * @return
+    */
+  def allCoordinates(wire: Wire): List[Coordinate] = {
+    val turningPoints: List[Coordinate] = findTurningPoints(wire)
+
+    val startingAndEndingPoints = wire.path
+      .zip(turningPoints.init)
+      .zip(turningPoints.tail)
+      .map(d => (d._1._1, StartingAndEndingPoints(d._1._2, d._2)))
+    startingAndEndingPoints.flatMap(d => pathTravelled(d._1, d._2))
+  }
+
+  /**
+    *
+    * @param wire
+    * @return
+    */
+  def findTurningPoints(wire: Wire): List[Coordinate] = {
+    wire.path
+      .foldLeft(List[Coordinate](Coordinate(0, 0)), Coordinate(0, 0))(
+        (accum, c) =>
```

Comment:
It's always best to match on tuples and give each element a better name.  Instead of just finding the turning points, you could find all the coordinates in a single fold:

```scala
    val origin = Coordinate(0, 0)
    (wire.path.foldLeft(List[Coordinate]() -> origin) {
      case ((acc, currentCoordinate), displacement) =>
        val path = coordinatesAlong(displacement, currentCoordinate)
        (acc ::: path, path.last)
    })._1
```

Where `coordinatesAlong` calculates all the coordinates for a given starting coordinate and displacement.  
What do you think?